File: cpp/src/p23_2/main.cc
#include <algorithm>
#include <iostream>
#include <numeric>
#include <thread>
#include <vector>

bool isAbundant(const int n) {
  // skip 1, count it in the sum, because the value itself isn't a divisor but 1 is.
  // This way, for all other divisors we can catch both at the same time (2 and n/2 for example)
  // and stop earlier
  int sum = 1;
  int stop = n-1;
  for (int i = 2; i < stop; ++i) {
    if (n % i == 0) {
      sum += i;
      stop = n/i;
      sum += stop != i ? stop : 0;
    }
  }
  return (sum > n);
}

bool isSumOfTwoAbundants(int n, const std::vector<int>& abundantNumbers) {
  const int abundant_size = static_cast<int>(abundantNumbers.size());
  for (int i = 0; i < abundant_size; ++i) {
    const int smallestSumForI = abundantNumbers[i] + abundantNumbers[i];
    if (smallestSumForI > n) {
      break;
    }

    for (int j = i; j < abundant_size; ++j) {
      const int sum = abundantNumbers[i] + abundantNumbers[j];
      if (sum == n) {
        return true;
      }

      if (sum > n) {
        break;
      }
    }
  }

  return false;
}

int main() {
  const int numThreads = std::thread::hardware_concurrency();

  // abundant numbers for thread data. Each threads puts results here.
  // numbers that are abundant get 1, others get 0
  std::vector<int> abundantNumbers(28123);

  // Phase 1: stride through the data and generate per-thead abundant numbers
  std::vector<std::thread> threads;
  for (int threadNum = 0; threadNum < numThreads; ++threadNum) {
    threads.emplace_back([&abundantNumbers, numThreads, threadNum]() {
      for (int i = 1+threadNum; i <= 28123; i += numThreads) {
        abundantNumbers[i-1] = isAbundant(i);
      }
    });
  }
  for (auto& t : threads) {
    t.join();
  }
  threads.clear();

  // Phase 2: Go back through and update the numbers "correctly".
  // Even though it doesn't take much space, we'll do it in-place
  int nextIndex = 0;
  for (int i = 0; i < 28123; ++i) {
    if (abundantNumbers[i] > 0) {
      abundantNumbers[nextIndex] = i+1;
      ++nextIndex;
    }
  }
  abundantNumbers.resize(nextIndex);

  // Phase 3: split back out again and calculate the per-thread sums
  std::vector<int> perThreadSums(numThreads, 0);
  for (int threadNum = 0; threadNum < numThreads; ++threadNum) {
    threads.emplace_back([&abundantNumbers, &perThreadSums, numThreads, threadNum]() {
      for (int i = 1+threadNum; i <= 28123; i += numThreads) {
        if (!isSumOfTwoAbundants(i, abundantNumbers)) {
          perThreadSums[threadNum] += i;
        }
      }
    });
  }
  for (auto& t : threads) {
    t.join();
  }
  threads.clear();

  // Phase 4: "merge" the per-thread sums
  const int sum = std::accumulate(perThreadSums.begin(), perThreadSums.end(), 0);
  std::cout << sum << "\n";
}

File: golang/src/p23_2/main.go
package main

import (
  "fmt"
  "runtime"
  "sync"
)

func main() {
  cpus := runtime.NumCPU()

  // First pass, the index will be 0 or 1 if abundant. Second pass, it'll be the list of abundant numbers.
  abundantNumbers := make([]int, 28123)

  // Use wait groups to sync tasks
  var abundantWG sync.WaitGroup
  abundantWG.Add(cpus)

  // Phase 1: stride through the data (earlier numbers are easier so striding helps balance it)
  // fill in 0 or 1 if the index is abundant or not
  for t := 0; t < cpus; t++ {
    go func(threadNum int) {
      for i := 1+threadNum; i <= 28123; i += cpus {
        if isAbundant(i) {
          abundantNumbers[i-1] = 1
        } else {
          abundantNumbers[i-1] = 0
        }
      }
      abundantWG.Done()
    }(t)
  }
  abundantWG.Wait()

  // Phase 2: In-place change abundantNumbers to the list of abundant numbers instead of 0/1 per index
  nextIndex := 0
  for i, num := range abundantNumbers {
    if num == 1 {
      abundantNumbers[nextIndex] = i+1
      nextIndex++
    }
  }
  abundantNumbers = abundantNumbers[0:nextIndex] // resize to the actual number of abundant numbers

  // Phase 3: split back out again and calculate per-thread sums
  var sumWG sync.WaitGroup
  sumWG.Add(cpus)

  sums := make([]int, cpus)
  for t := 0; t < cpus; t++ {
    go func(threadNum int) {
      for i := 1+threadNum; i <= 28123; i += cpus {
        if !isSumOfTwoAbundants(i, abundantNumbers) {
          sums[threadNum] += i
        }
      }
      sumWG.Done()
    }(t)
  }
  sumWG.Wait()

  // Phase 4: "merge" the per-thread sums
  sum := 0
  for _, s := range sums {
    sum += s
  }

  fmt.Println(sum)
}

func isAbundant(n int) bool {
  sum := 1
  stop := n-1
  for i := 2; i < stop; i++ {
    if n % i == 0 {
      sum += i
      stop = n/i
      if stop != i {
        sum += stop
      }
    }
  }
  return sum > n
}

func isSumOfTwoAbundants(n int, abundantNumbers []int) bool {
  for i, abundantX := range abundantNumbers {
    smallestSum := abundantX + abundantX
    if smallestSum > n {
      break;
    }
    for j := i; j < len(abundantNumbers); j++ {
      sum := abundantX + abundantNumbers[j]
      if sum == n {
        return true
      }
      if sum > n {
        break
      }
    }
  }

  return false
}
